lib/ansible/module_utils/aws/core.py

```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
         else:
             self._module.fail_json(msg=message, exception=last_traceback,
                                    **camel_dict_to_snake_dict(response))
+
+    def create_response_dict(self, item):
+        if isinstance(item, string_types):
+            return item
+        elif isinstance(item, list):
+            return [self.create_response_dict(i) for i in item]
+        # from collections import OrderedDict fails on Python 2.6
+        # botocore uses its own OrderedDict class that is compatible with Python 2.6
+        elif item.__class__.__name__ == 'OrderedDict':
+            item = dict(item)
+            for k, v in item.items():
+                item[k] = self.create_response_dict(v)
+        return item
+
+    def populate_response(self, response, extra_output):
+        if isinstance(response, string_types):
+            return response
+        elif isinstance(response, list):
+            return [self.populate_response(i, extra_output) for i in response]
+        elif isinstance(response, dict):
+            for output_key, v in dict(response).items():
+                if output_key in extra_output:
+                    response[output_key] = extra_output[output_key]
+                else:
+                    # Careful. If doing diffs with deeply nested structures with repeat keys this will
+                    # do weird things. Safer to pass extra_output as a complex structure in cases like that.
+                    # TODO: make extra_output smarter
+                    response[output_key] = self.populate_response(v, extra_output)
+        return response
+
+    def call_with_check_mode(self, client, method, extra_output, params):
+
+        arg_gen = ArgumentGenerator()
+        resp_stub = Stubber(client)
+        op = client.meta.method_to_api_mapping.get(method)
+        output_shape = client._service_model.operation_model(op).output_shape
+        if not output_shape:
+            return None
+        output_args = arg_gen.generate_skeleton(output_shape)
+
+        # Generate response dict
+        resp = self.create_response_dict(output_args)
+        if extra_output:
+            resp = self.populate_response(dict(resp), extra_output)
+
+        # Validate input parameters
+        resp_stub.add_response(method, resp)
+        with resp_stub:
+            return getattr(client, method)(**params)
+
+    def get_resource_before(self, client, method, params, output_dict=None):
+        if output_dict is not None:
+            return output_dict
+        if method:
+            if not any(method.startswith(r) for r in ('get', 'describe', 'list')):
+                self.warn("This may be an implementation error of diff mode. The compare "
+                          "method should be a read-only method. Setting 'before' to an empty dict.")
```

Comment:
Yeah, I wish we had better metadata around this. You can look at the http method, but POST might be used for read-only stuff for some services. I'm *pretty sure* GET methods don't modify state, but not 100%.